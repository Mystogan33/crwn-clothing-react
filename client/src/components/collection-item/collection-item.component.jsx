import React from 'react';
import { connect } from 'react-redux';

import { addItem } from '../../redux/cart/cart.actions';

import {
  CollectionItemContainer,
  AddButton,
  BackgroundImage,
  CollectionFooterContainer,
  NameContainer,
  PriceContainer
} from './collection-item.styles';

const mapDispatchToProps = dispatch => ({
  addItem: item => dispatch(addItem(item))
})

export const CollectionItem = connect(null, mapDispatchToProps)(
  ({item, addItem}) => {
    const { name, price, imageUrl } = item;
    return (
      <CollectionItemContainer>
        <BackgroundImage className='image' imageUrl={imageUrl} />
        <CollectionFooterContainer>
          <NameContainer>{name}</NameContainer>
          <PriceContainer>${price}</PriceContainer>
        </CollectionFooterContainer>
        <AddButton onClick={() => addItem(item)} inverted>Add to cart</AddButton>
      </CollectionItemContainer>
    )
  }
);
