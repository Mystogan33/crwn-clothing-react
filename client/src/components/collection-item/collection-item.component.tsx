import React, { FC } from 'react';
import { ICollectionItem } from '../../interfaces/interfaces';
import { addItem } from '../../redux/cart/cartSlice';
import { useAppDispatch } from '../../redux/hooks';

import {
  CollectionItemContainer,
  AddButton,
  BackgroundImage,
  CollectionFooterContainer,
  NameContainer,
  PriceContainer
} from './collection-item.styles';

type CollectionItemProps = {
  item: ICollectionItem
};

export const CollectionItem: FC<CollectionItemProps> = ({ item }) => {
  const { name, price, imageUrl } = item;
  const dispatch = useAppDispatch();
  const addItemCart = () => dispatch(addItem(item));

  return (
    <CollectionItemContainer>
      <BackgroundImage className='image' imageUrl={imageUrl} />
      <CollectionFooterContainer>
        <NameContainer>{name}</NameContainer>
        <PriceContainer>${price}</PriceContainer>
      </CollectionFooterContainer>
      <AddButton onClick={addItemCart} inverted>Add to cart</AddButton>
    </CollectionItemContainer>
  );
};
