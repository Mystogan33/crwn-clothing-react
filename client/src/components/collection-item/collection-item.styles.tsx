import styled from 'styled-components';
import { CustomButton }  from '../custom-button/custom-button.component';

const CollectionItemContainer = styled.div`
  width: 22vw;
  display: flex;
  flex-direction: column;
  height: 350px;
  align-items: center;
  position: relative;
  transition: opacity .5s ease-in-out, box-shadow .4 ease-in-out;

  &:hover {
    .image {
      opacity: 0.8;
      box-shadow: 0 1px 10px rgba(10, 10, 10, 0.7);
    }
    button {
      opacity: 0.85;
      display: flex;
    }
  }

  @media screen and (max-width: 800px) {
    width: 40vw;

    &:hover {
      .image, button {
        opacity: unset;
      }
    }
  }
`;

const AddButton = styled(CustomButton)`
  width: 80%;
  opacity: 0.7;
  position: absolute;
  top: 255px;
  display: none;
  border-radius: 0.2rem;

  @media screen and (max-width: 800px) {
    display: block;
    opacity: 0.85;
    min-width: unset;
    padding: 0 10px;
  }
`;

 const BackgroundImage = styled.div<{ imageUrl? : string}>`
  width: 100%;
  height: 95%;
  background-size: cover;
  background-position: center;
  margin-bottom: 5px;
  border-radius: 0.3rem;
  background-image: ${({ imageUrl }) => `url(${imageUrl})`};
  transition: all .3s ease-in-out;
  box-shadow: 0 1px 5px rgba(10, 10, 10, 0.35);
`;

 const CollectionFooterContainer = styled.div`
  width: 100%;
  height: 5%;
  display: flex;
  justify-content: space-between;
  font-size: 18px;
`;

 const NameContainer = styled.span`
  width: 90%;
  margin-bottom: 15px;
`;

const PriceContainer = styled.span`
  width: 10%;
  text-align: right;
`;

export {
  CollectionItemContainer,
  AddButton,
  BackgroundImage,
  CollectionFooterContainer,
  NameContainer,
  PriceContainer
};
