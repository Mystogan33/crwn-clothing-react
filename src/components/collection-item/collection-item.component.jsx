import React from 'react';
import './collection-item.styles.scss';

import { CustomButton } from '../index';

export const CollectionItem = ({id, name, price, imageUrl}) => (
  <div className="collection-item">
    <div
      className="image"
      style={{
        backgroundImage: `url(${imageUrl})`
      }} />
    <div className="collection-footer">
      <span className="name">{name}</span>
      <span className="price">{price}</span>
    </div>
    <CustomButton inverted>Add to cart</CustomButton>
  </div>
);
