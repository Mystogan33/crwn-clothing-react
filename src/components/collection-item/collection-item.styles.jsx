import styled from 'styled-components';
import { CustomButton }  from '../custom-button/custom-button.component';

const CollectionItemContainer = styled.div`
  width: 22vw;
  display: flex;
  flex-direction: column;
  height: 350px;
  align-items: center;
  position: relative;

  &:hover {
    .image {
      opacity: 0.8;
      box-shadow: 0 1px 10px rgba(10, 10, 10, 0.7);
    }
    button {
      opacity: 0.85;
      display: flex;
    }
  }
`;

const AddButton = styled(CustomButton)`
  width: 80%;
  opacity: 0.7;
  position: absolute;
  top: 255px;
  display: none;
  border-radius: 0.2rem;
`;

 const BackgroundImage = styled.div`
  width: 100%;
  height: 95%;
  background-size: cover;
  background-position: center;
  margin-bottom: 5px;
  background-image: ${({ imageUrl }) => `url(${imageUrl})`};
  border-radius: 0.3rem;
  transition: opacity .3s ease-in-out, box-shadow .4s ease-in-out;
`;

 const CollectionFooterContainer = styled.div`
  width: 100%;
  height: 5%;
  display: flex;
  justify-content: space-between;
  font-size: 18px;
`;

 const NameContainer = styled.span`
  width: 90%;
  margin-bottom: 15px;
`;

const PriceContainer = styled.span`
  width: 10%;
  text-align: right;
`;

export {
  CollectionItemContainer,
  AddButton,
  BackgroundImage,
  CollectionFooterContainer,
  NameContainer,
  PriceContainer
};
