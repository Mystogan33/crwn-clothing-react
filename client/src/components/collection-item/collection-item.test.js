import React from 'react';
import { shallow } from 'enzyme';

import { CollectionItem } from './collection-item.component';
import { BackgroundImage, CollectionFooterContainer, NameContainer, PriceContainer, AddButton } from './collection-item.styles';

describe('Collection Item Component', () => {
  let wrapper;
  let mockItem;
  let mockProps;
  let mockAddItem;

  beforeEach(() => {

    mockItem = {
      id: 999,
      imageUrl: "https://testimage.com",
      name: "Mock item",
      price: 25,
      quantity: 1
    };

    mockAddItem = jest.fn();

    mockProps = {
      item: mockItem,
      addItem: mockAddItem
    };

    wrapper = shallow(<CollectionItem {...mockProps} />);
  });

  it('should render Collection Item component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should properly render BackgroundImage and render it with imageUrl and className equal to "image"', () => {
    expect(wrapper.exists(BackgroundImage)).toBe(true);
    expect(wrapper.find(BackgroundImage).prop('className')).toEqual('image');
    expect(wrapper.find(BackgroundImage).prop('imageUrl')).toEqual(mockItem.imageUrl);
  });

  it('should properly render CollectionFooterContainer', () => {
    expect(wrapper.exists(CollectionFooterContainer)).toBe(true);
    expect(wrapper.find(CollectionFooterContainer).exists(NameContainer)).toBe(true);
    expect(wrapper.find(CollectionFooterContainer).exists(PriceContainer)).toBe(true);
  });

  it('should render NameContainer with item name', () => {
    expect(wrapper.find(NameContainer).text()).toBe(mockItem.name);
  });

  it('should render PriceContainer with item price', () => {
    expect(wrapper.find(PriceContainer).text()).toBe(`$${mockItem.price.toString()}`);
  });

  it('should properly render AddButton', () => {
    expect(wrapper.exists(AddButton)).toBe(true);
  });

  it('should trigger mockAddItem when AddButton is clicked', () => {
    wrapper.find(AddButton).simulate('click');
    expect(mockAddItem).toHaveBeenCalledWith(mockItem);
  });
});