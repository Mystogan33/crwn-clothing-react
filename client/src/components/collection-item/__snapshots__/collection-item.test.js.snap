// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection Item Component should render Collection Item component 1`] = `ShallowWrapper {}`;
